Refactor Engine Strategies & Add Prolog Engine Strategy

alaeddine-66 opened the pull request:
# Refactor Engine Strategies & Add Prolog Engine Strategy

### Added
- `PrologEngineStrategy` for puzzle-solving and `EngineStrategyFactory` for managing engine strategies

### Changes
- Refactored `EngineStrategy` and `LogicAgent` 
- Updated CBMC and Z3 strategies for compatibility
- Tests for Prolog engine in `test_prolog_engine_strategy.py`

### Testing
```bash
python -m unittest agent/logic/tests/test_prolog_engine_strategy.py
python -m unittest agent/logic/tests/test_zebra_benchmark.py

alaeddine-66 committed:
Refactor engine strategies and add PrologEngineStrategy

pkesseli commented:
> Thanks for the update! Good luck for finishing your paper, looking forward to your review

Hi @alaeddine-66 , just letting you know I haven't forgotten about this PR! It is just that I will be switching fro Meta to Microsoft AI in January, and thus need to set up Polymath in a new inference environment. So apologies for the delay in review, as soon as I have access to the new inference compute, I will review and approve asap!

pkesseli commented:
Hi @alaeddine-66, thanks for your patience! I finally have access again. Overall looks good and granular, thank you, just a few comments:

1. Please check all files with `pyright`, e.g. `prolog_engine_strategy.py` still has some pyright/type errors:
```
/polymath/agent/logic/prolog_engine_strategy.py
  /polymath/agent/logic/prolog_engine_strategy.py:166:32 - error: Type "List[str] | None" is not assignable to declared type "list[str]"
    Type "List[str] | None" is not assignable to type "list[str]"
      "None" is not assignable to "list[str]" (reportAssignmentType)
  /agent/logic/prolog_engine_strategy.py:167:23 - error: Type "int | None" is not assignable to declared type "int"
    Type "int | None" is not assignable to type "int"
      "None" is not assignable to "int" (reportAssignmentType)
2 errors, 0 warnings, 0 informations
```
2. `agent.py`: `solverSpec.content = stdout` mutates the input object, reusing it to mean two different things (constraints, then solver output). Keep SolverConstraints immutable and pass `stdout` as a separate parameter to `parse_solver_output`.
3. `engine_strategy_factory.py`, `prolog_engine_strategy.py`, `test_prolog_engine_strategy.py`are missing the Meta copyright header.

alaeddine-66 committed:
Merge branch 'facebookresearch:main' into pr-engine-refactor

alaeddine-66 committed:
Fix Pyright errors and add Meta headers

alaeddine-66 commented:
Hi @pkesseli,

Thanks for the review! I've just pushed a commit addressing all your feedback.

As a quick side note from my local testing (using LLaMA 3 70B quantized in 4-bit): the Prolog engine yields the exact same results as the CBMC approach but runs about 2x faster, because the generated DSL is much shorter and easier for the LLM to output.

Also, Prolog could be a great fit to evaluate the FOLIO benchmark down the line, since it natively supports logical operations like implication and iff.
